logs: handle dataplane-compatible logs-frames

gabor opened the pull request:
we update the logs-visualization code to be able to read dataframes that are created based on the dataplane spec : https://grafana.github.io/dataplane/logs
(we will adjust Loki to optionally output this kind of data in a future PR)

how to test:
1. make sure normal things still work (for example, run some Loki logql queries etc)
2. use the test-datasource (`gdev-testdata`), choose the `raw frames` mode, and paste in the following JSON:
```json
[
  {
    "schema": {
      "refId": "A",
      "meta": {
        "type": "log-lines",
        "preferredVisualisationType": "logs"
      },
      "fields": [
        { "name": "timestamp", "type": "time" },
        { "name": "body", "type": "string" },
        { "name": "severity", "type": "string" },
        { "name": "attributes", "type": "other" },
        { "name": "id", "type": "string" }
      ]
    },
    "data": {
      "values": [
        [1686648201526, 1686648200232, 1686648200714, 1686648200008, 1686648199345, 1686648198480, 1686648198873],
        ["line1", "line2", "line3", "line4", "line5", "line6", "line7"],
        ["info", "warning", "error", "error", "warning", "error", "info"],
        [
          { "source": "client", "place": "luna1", "nested": ["one", "two", "three"] },
          { "source": "server", "place": "luna2" },
          { "source": "server", "place": "luna1" },
          { "source": "client", "place": "luna2" },
          { "source": "client", "place": "luna1" },
          { "source": "server", "place": "luna2" },
          { "source": "client", "place": "luna1" }
        ],
        ["id1", "id2", "id3", "id4", "id5", "id6", "id7"]
      ],
      "nanos": [[52946, 0, 531000, 531000, 955000, 0, 0], null, null, null, null]
    }
  }
]
```

make sure you see logs, they have the correct log-level-color etc.

some features are still missing, will be added in later PR to keep the diffs small:
1. the dataplane-contract uses the timefield's native nanosecond support to handle nanosecond-precision timestamps (opposed to how the loki-style dataframes have a `tsNs` string-field with timestamps-as-strings). the support for this will come in a separate PR.
2. in logs-details currently we hide certain attributes, like timestamp&id.. we also hide`labels`. now we have `attributes`, so we should hide that one when it is correct to hide it  (probably switch based on the dataframe-type?)
3. the dataplane-contract uses `attributes`, not `labels`. these are basically labels, but the values can be any structures, not just `string`..they can be "nested". for now, we just `.toString()` them in `logsFrame.ts`, in the future we should have full support for them. (this may be a long-term plan)

( part of https://github.com/grafana/grafana/issues/67657 )

gabor committed:
logs: handle dataplane-compatible logs-frames

svennergr commented on .betterer.results:
```diff
@@ -2768,6 +2768,9 @@ exports[`better eslint`] = {
     "public/app/features/live/index.ts:5381": [
       [0, 0, 0, "Do not use any type assertions.", "0"]
     ],
+    "public/app/features/logs/logsFrame.ts:5381": [
+      [0, 0, 0, "Unexpected any. Specify a different type.", "0"]
+    ],
```
I think this is a relict and might need a `yarn betterer --update`?

gabor replied:
good catch. done!

svennergr commented on public/app/core/logsModel.ts:
```diff
@@ -359,39 +337,27 @@ export function logSeriesToLogsModel(logSeries: DataFrame[], queries: DataQuery[
-        const labels = getAllLabels(info);
-        if (labels.length > 0) {
-          allLabels.push(labels);
+        if (frameLabels != null && frameLabels.length > 0) {
```
I think based on https://github.com/grafana/grafana/pull/70318/files#diff-07da85e1ede94286e302d4831fd8768e312dd25291a32fb9bfb1f02115331121R50 `frameLabels` can also be `undefined`? I think it's better to check agains `!frameLabels` anyways.

gabor replied:
this will be a long discussion :) 

basically, `x != null` is the same as `(x !== null) && (x !== undefined)`. i nearly always write `!= null`, regardless of whether i am comparing for `null` or `undefined`. it's a trick, and maybe tricks are bad.

to clarify, is your recommendation to go with
```js
if (frameLabels && frameLabels.length > 0)
```
?

svennergr commented on public/app/features/logs/legacyLogsFrame.ts:
```diff
@@ -0,0 +1,50 @@
+  const labelsField = cache.getFieldByName('labels');
+
+  if (labelsField == null) {
+    return undefined;
```
In my opinion we should rather return `null` and not `undefined`.

svennergr commented on public/app/features/logs/legacyLogsFrame.ts:
```diff
@@ -0,0 +1,50 @@
+
+export function parseLegacyLogsFrame(frame: DataFrame): LogsFrame | null {
+  const cache = new FieldCache(frame);
+  const timeField = cache.getFields(FieldType.time)[0];
+  const bodyField = cache.getFields(FieldType.string)[0];
```
Is this error safe? What if `cache.getFields(FieldType.string)` returns an empty array?

gabor replied:
it is checked on the next line:
```js
  // these two are mandatory
  if (timeField === undefined || bodyField === undefined) {
    return null;
  }
```

if `getFields` returns an empty-array, then any item from it will be `undefined`.


would you prefer the checking to happen more explicitly?

like:
```ts
const timeFields = cache.getFields(FieldType.time);
const bodyFields = cache.getFields(FieldType.string);

if (timeFields.length === 0 || bodyFields.length === 0) {
    return null;
}

const timeField = timeFields[0];
const bodyField = bodyFields[0];
```

?

svennergr commented on public/app/features/logs/logsFrame.test.ts:
```diff
@@ -0,0 +1,167 @@
+    });
+
+    expect(result).not.toBeNull();
+    if (result == null) {
+      // to make typescript happy
+      throw new Error('should never happen');
+    }
```
You can also use the non-null assertion operator `!`.

gabor replied:
i hate that operator (details on demand), but you're right, in a test it should be ok. will change it.

gabor replied:
done.

svennergr commented on public/app/features/logs/logsFrame.ts:
```diff
@@ -0,0 +1,68 @@
+
+function getField(cache: FieldCache, name: string, fieldType: FieldType): FieldWithIndex | undefined {
+  const field = cache.getFieldByName(name);
+  if (field == null) {
+    return undefined;
+  }
```
Also return `null` here? What would be the difference to https://github.com/grafana/grafana/pull/70318/files#diff-07da85e1ede94286e302d4831fd8768e312dd25291a32fb9bfb1f02115331121R42-R44?

gabor replied:
hmm.. for now i changed it to be more regular to:
```js
if (field === undefined) {
    return undefined;
}
```

gabor replied:
and opene a discussion at the bottom.

gabor commented:
i adjusted the code slightly, i added a `getAttributesAsLogs` function to the `LogsFrame` structure. i'm planning to remove it later, but for now it helps us while we did not completely migrate to `attributes`.

i wanted to avoid the process-all-labels code, for situations where we know the attributes are simple.

gabor committed:
simpler tests

gabor committed:
simpler code

gabor committed:
simpler code

gabor commented:
@svennergr thanks for the review, i'll respond to https://github.com/grafana/grafana/pull/70318#discussion_r1234940276 and https://github.com/grafana/grafana/pull/70318#discussion_r1234936470 here, because they're about the same topic...

all the code in `logsFrame.ts` and `logsFrameLegacy.ts` is basically calling a bunch of `FieldCache.getFields` and then populating the `LogsFrame` structure).

now:
1. `FieldCache.getFields` returns data or `undefined`
2. currently, in `type LogsFrame`, i marked optional stuff as `something?`, which translates to `something|undefined`. 

i see you'd prefer to move more to `null`s for optionality, correct?

so would you change `type LogsFrame` to be like:
```ts
export type LogsFrame = {
  timeField: FieldWithIndex;
  bodyField: FieldWithIndex;
  timeNanosecondField: FieldWithIndex|null;
  severityField: FieldWithIndex|null;
  idField: FieldWithIndex|null;
  attributes: Attributes[]|null;
}
```

?

svennergr commented:
> i see you'd prefer to move more to `null`s for optionality, correct?

Hm, I think that's more about the old discussion about when to use `undefined` or `null`? 
E.g. in cases where we do 

```
  if (field == null) {
    return undefined;
  }
```

we deliberately make the decision to return something. And I would always return `null` in those cases to show the absence of a true-ish/real return value.

But I also see that this could be troublesome together with optional arguments.

gabor committed:
switch to null from undefined

gabor committed:
updated tests

gabor committed:
added test for simplest case

gabor committed:
more direct comparisons

gabor committed:
simplified code

svennergr commented on public/app/features/logs/logsFrame.ts:
```diff
@@ -0,0 +1,87 @@
+  return field.type === fieldType ? field : undefined;
+}
+
+const z: undefined | number = undefined;
+
+if (z ?? 0 > 0) {
+  console.log('yo');
+}
```
yo, do we need this? 😅

```suggestion
```

gabor replied:
good catch, removed! 😂

svennergr reviewed (approved):
Just a nit, rest LGTM.

gabor committed:
removed by-mistake added code
